src/VisualStudio/Core/Def/Implementation/TableDataSource/Suppression/VisualStudioSuppressionFixService.cs

```diff
@@ -0,0 +1,295 @@
+            foreach (var diagnosticData in diagnosticsToFix)
+            {
+                List<DiagnosticData> diagnosticsPerDocument;
+                if (!builder.TryGetValue(diagnosticData.DocumentId, out diagnosticsPerDocument))
+                {
+                    diagnosticsPerDocument = new List<DiagnosticData>();
+                    builder[diagnosticData.DocumentId] = diagnosticsPerDocument;
+                }
+
+                diagnosticsPerDocument.Add(diagnosticData);
+            }
+
+            if (builder.Count == 0)
+            {
+                return ImmutableDictionary<Document, ImmutableArray<Diagnostic>>.Empty;
+            }
+
+            var finalBuilder = ImmutableDictionary.CreateBuilder<Document, ImmutableArray<Diagnostic>>();
+            var latestDocumentDiagnosticsMap = new Dictionary<DocumentId, ImmutableHashSet<DiagnosticData>>();
+            foreach (var group in builder.GroupBy(kvp => kvp.Key.ProjectId).AsImmutable())
+            {
+                var projectId = group.Key;
+                var project = _workspace.CurrentSolution.GetProject(projectId);
+                if (project == null || !shouldFixInProject(project))
+                {
+                    continue;
+                }
+
+                var uniqueDiagnosticIds = group.SelectMany(kvp => kvp.Value.Select(d => d.Id)).ToImmutableHashSet();
+                var latestProjectDiagnostics = (await _diagnosticService.GetDiagnosticsForIdsAsync(project.Solution, project.Id, diagnosticIds: uniqueDiagnosticIds, includeSuppressedDiagnostics: true, cancellationToken: cancellationToken)
+                    .ConfigureAwait(false)).Where(d => d.DocumentId != null);
+
+                latestDocumentDiagnosticsMap.Clear();
+                foreach (var kvp in latestProjectDiagnostics.GroupBy(d => d.DocumentId))
+                {
+                    latestDocumentDiagnosticsMap.Add(kvp.Key, kvp.ToImmutableHashSet());
+                }
+
+                var documentsToTreeMap = await GetDocumentIdsToTreeMapAsync(project, cancellationToken).ConfigureAwait(false);
+                foreach (var documentDiagnostics in group)
+                {
+                    var document = project.GetDocument(documentDiagnostics.Key);
+                    if (document == null)
+                    {
+                        continue;
+                    }
+
+                    ImmutableHashSet<DiagnosticData> latestDocumentDiagnostics;
+                    if (!latestDocumentDiagnosticsMap.TryGetValue(document.Id, out latestDocumentDiagnostics))
+                    {
+                        // Ignore stale diagnostics in error list.
+                        continue;
+                    }
+
+                    // Filter out stale diagnostics in error list.
+                    var documentDiagnosticsToFix = documentDiagnostics.Value.Where(d => latestDocumentDiagnostics.Contains(d));
```

Comment:
@heejaechang I have addressed your feedback, and now query the diagnostic service to fetch upto-date diagnostics and filter out stale ones in error list. I'll test it on some big solution later today and see how it goes.